### PR TITLE
Improve the bureaucracy system

### PR DESCRIPTION
**Pull request by twilightwanderer**
## About The Pull Request

Adding a blank printing system 
large_stamp-hop

## Why It's Good For The Game

Virtually every Space Station 13 server has its own system of bureaucracy. As a rule, it is some kind of Wiki page with markup tags and a few sample forms. And that's usually the end of it. With stamps and a filing system in place, people rarely use them. This PR takes it upon itself to remedy this situation by introducing form standards into the game itself and adding the ability to print them.

i5KkoLbRql

The printing system presented in this PR is based on the Paradise Station SS220 server. On this server, the system of printing forms exists for several years and has proved itself perfectly (queues to the head of the staff of 40 assistants will not let you lie). 

I plan to transfer a similar system to TGStation by creating a database of standard forms and making it possible to print them. 

The forms currently have the following system: 
Forms are json objects placed on **config/blanks.json**. It contains four variables:
- **code** - The form code displayed in the list of forms. (Hello to the movie Brazil). Made to create an element of mechanics and role-playing in the style of, "Go into the brig to get a form like NT-SCR-EO, present it to the head doctor to get an NT-MDC-DC form, then come back with that form to me for a 5 credit award certificate." I'm exaggerating all this, of course; 
- **category** - The prenativity of the form according to the department. 
- **name** - The normal name of the form displayed on it;
- **info** - Form content. The content itself is html markup.

These forms are available for printing in any printer.

## Discussions of issues

I suggest using forms like this (I look forward to your suggestions):

_Legend:_
 `Developed and approved`
 `Developed and awaiting review`
 `There are comments, in the process of correcting`
 `In the process of developing`
 `Will not be entered in this PR`

**Command**
 `NT-CMD-JA Job application`

<details>

изображение

изображение

</details>

 `NT-CMD-JT Job termination`

<details>

изображение

</details>

 `NT-CMD-JR Job reassignment`

<details>

изображение

</details>

 `NT-CMD-C Complaint form`

<details>

изображение

</details>

 `NT-CMD-O Order form`

<details>

изображение

</details>

 `NT-CMD-PH Permit to perform hazardous work`

<details>

изображение

</details>

**Security**
 `NT-SCR-W Warrant`

<details>

изображение

</details>

 `NT-SCR-SW Search warrant`

<details>

изображение

</details>

 `NT-SCR-IR Interrogation Report`

<details>

изображение

изображение

</details>

 `NT-SCR-SR Security Report` 

<details>

изображение

</details>

 `NT-SCR-TR Trial report`
 `NT-SCR-EO Execution order`

<details>

изображение

</details>

 `NT-SCR-PP Permission to possess`

<details>

изображение

</details>

**Engineering**
 `NT-ENG-PW Permission to perform non-standard work`

<details>

изображение

</details>

 `NT-ENG-PE Permit for the issuance of engineering equipment`

<details>

изображение

изображение

</details>

**Research**
 `NT-RSH-PR Permission to perform non-standard research`

<details>

изображение

</details>

 `NT-RSH-PE Permit for the issuance of research equipment`

<details>

изображение

изображение

</details>

**Medical**
 `NT-MDC-MR Medical Record`
 `NT-MDC-MP Medical Prescription`

<details>

изображение

</details>

 `NT-MDC-AR Autopsy Report`

<details>

изображение

</details>

 `NT-MDC-DC Death Certificate`

<details>

изображение

</details>

**Supply**
 `NT-SPL-PO Permission to execute a non-standard order`

<details>

изображение

изображение

</details>

**Other**

 `NT-LB Lunch bill`

<details>

изображение

</details>

 `NT-PS Provision of paid services`

<details>

изображение

</details>

Ability to use graphics in the forms. 

The  tag allows you to use all possible graphics to increase interactivity. You can put logos, maps and other things in it. 

изображение

## Bags 

Bugs discovered during the development process that jeopardize the implementation of the project.

The stamp image is not dragged when you scroll.

<details>

zof4eCkYzD

</details>

## Changelog

:cl:
add: Added a system of forms and the ability to print them
/:cl:

### CODE REVIEW

File: code/modules/paperwork/paper.dm
Comment (Fikou) [inline code/modules/paperwork/paper.dm:400]: dont document it on the objects, document the vars themselves

File: code/modules/paperwork/paper.dm
Comment (Fikou) [inline code/modules/paperwork/paper.dm:408]: make the categories defines so you cant fuck up by doing a typo

File: code/modules/paperwork/paper.dm
Comment (Fikou) [inline code/modules/paperwork/paper.dm:409]: capitalize both words?

File: code/modules/paperwork/photocopier.dm
Comment (Fikou) [inline code/modules/paperwork/photocopier.dm:67]: here too

File: code/modules/paperwork/paper.dm
Comment (twilightwanderer) [reply code/modules/paperwork/paper.dm:408]: > make the categories defines so you cant fuck up by doing a typo

What's the best way to make them? Through #define?

File: code/modules/paperwork/photocopier.dm
Comment (twilightwanderer) [reply code/modules/paperwork/photocopier.dm:66]: > use descriptive var names

Do you mean listPaperBlanks?

Comment (twilightwanderer) [reply code/modules/paperwork/photocopier.dm:66]: I don't quite understand how to describe the name of the variable, if its task is only to generate and pass a list of all forms.

**Comment by oklews** [comment]
FUCK yeah, finally I'll feel like I can write warrants up as a HoS

**Comment by Valorium-RB** [comment]
Oh this is actually quite similar to the paperwork database I suggested a little while ago https://tgstation13.org/phpBB/viewtopic.php?f=9&t=29594
+1 for that

File: code/modules/paperwork/paper.dm
Comment (twilightwanderer) [reply code/modules/paperwork/paper.dm]: I redid the code according to your recommendation through the category definitions.

File: code/modules/paperwork/photocopier.dm
Comment (Fikou) [reply code/modules/paperwork/photocopier.dm:66]: var/blank_type in subtypesof

File: code/modules/paperwork/photocopier.dm
Comment (Fikou) [reply code/modules/paperwork/photocopier.dm:67]: blank_paper or whatever

File: code/modules/paperwork/paper.dm
Comment (Fikou) [reply code/modules/paperwork/paper.dm]: dont make datums for that yes, just defines is fine

Comment (twilightwanderer) [reply code/modules/paperwork/paper.dm]: Okay, I already did that.

**Comment by Sylphet** [comment]
Medical could use a form to fill out as a record of physical exams.

**Comment by twilightwanderer** [comment]
> Medical could use a form to fill out as a record of physical exams.

Thank you. I was thinking about this form, as well as the security records form. I guess I'll add them, but I still want to create forms that players would actually use. Unfortunately with the current tgstation gameplay, I don't see any doctor filling out paperwork for every patient. Maybe if only on some HRP server, maybe for that I will make this form.

File: code/modules/paperwork/photocopier.dm
Comment (tralezab) [inline code/modules/paperwork/photocopier.dm:67]: And then rename all the variables that use `B` and `bb` into `paper_type`

```suggestion
	for(var/obj/item/paper/blank/paper_type as anything in subtypesof(/obj/item/paper/blank))
```

File: code/modules/paperwork/paper.dm
Comment (tralezab) [inline code/modules/paperwork/paper.dm:19]: move all of these into a new file in the _DEFINES folder

File: code/modules/paperwork/paper.dm
Comment (tralezab) [inline code/modules/paperwork/paper.dm:481]: since they will be in a _DEFINES folder you will no longer need to undef these

**Review by tralezab — CHANGES_REQUESTED** [review]

File: code/modules/paperwork/paper.dm
Comment (tralezab) [reply code/modules/paperwork/paper.dm:19]: p.s. i'm saying this because there are 13 of them at this point, some defines in paperbin, some defines in photocopier, and so on.

**Comment by twilightwanderer** [comment]
May i ask why my PR was transferred to draft?

**Comment by DrDiasyl** [comment]
> May i ask why my PR was transferred to draft?

I guess because your forms are "Awaiting review"

**Comment by twilightwanderer** [comment]
> > May i ask why my PR was transferred to draft?
>
> I guess because your forms are "Awaiting review"

I just made this list in the hope that someone would review the forms and, if anything, help me correct them. English is just not my native language.

**Comment by DrDiasyl** [comment]
@Cyberboss so why this PR was marked as draft?

**Comment by DrDiasyl** [comment]
> > > May i ask why my PR was transferred to draft?
> >
> > I guess because your forms are "Awaiting review"
> 
> I just made this list in the hope that someone would review the forms and, if anything, help me correct them. English is just not my native language. 
> 

Also, why some forms was deleted? Like trial and medical record

**Comment by DrDiasyl** [comment]
And also yes @twilightwanderer, can you send those forms in-game screens? I can review them.

**Comment by twilightwanderer** [comment]
@DrDiasyl I have added pictures of each form to the description. I hope I don't upset anyone with my disappointed expectations.

**Comment by twilightwanderer** [comment]
My PR is waiting for a review. Again.

**Comment by twilightwanderer** [comment]
> 
> 
> > > > May i ask why my PR was transferred to draft?
> > > 
> > > 
> > > I guess because your forms are "Awaiting review"
> > 
> > 
> > I just made this list in the hope that someone would review the forms and, if anything, help me correct them. English is just not my native language.
> 
> Also, why some forms was deleted? Like trial and medical record

Honestly, I don't know how to apply them. Trial Report seems to me too boring and bulky for people to want to fill it out. And a medical history can be filled out on a console.

**Review by stylemistake — APPROVED** [review]
Approving with suggestions.

**Comment by twilightwanderer** [comment]
изображение

**Comment by twilightwanderer** [comment]
And so, I'm back in touch. I changed the way I store the forms. As I wrote earlier, I believe that such a method will make life easier for maintainers and server hosts. Each server will be able to use its own bureaucracy system without changing the tgstation build code. As for the forms I generated, let them remain the default forms to be used.

**Review by stylemistake — APPROVED** [review]

**Review by tralezab — CHANGES_REQUESTED** [review]
this pr has been waitin' around awhile and I do like it a good bit so ping me on discord when you've handled the review so I know when to look at it again: `armhulenn#9821`

File: code/modules/paperwork/photocopier.dm
Comment (twilightwanderer) [reply code/modules/paperwork/photocopier.dm:65]: That's what I would have done. Just with TRY-CATH I check not only if the file exists, but also if the file itself is empty and catch errors in decoding, if something is wrong with the config.

**Comment by MrDoomBringer** [comment]
if this ever gets merged i want to make these able to be player-saved in the library console

**Comment by itseasytosee** [comment]
Uh oh one day left

**Comment by twilightwanderer** [comment]
> Uh oh one day left

I don't think my PR will ever be accepted, so I decided to stop developing it. Sorry about that.

**Comment by itseasytosee** [comment]
> > Uh oh one day left
> 
> I don't think my PR will ever be accepted, so I decided to stop developing it. Sorry about that.

image

File: tgui/packages/tgui/interfaces/Photocopier.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Photocopier.js:47]: These commas seem to only clutter the UI (they are rendered as text, yes, they can be safely omitted).

**Review by stylemistake — COMMENTED** [review]
I have no real objections against this PR so I might as well merge it. I took the liberty to remove commas from the UI.

File: code/modules/paperwork/photocopier.dm
Comment (stylemistake) [reply code/modules/paperwork/photocopier.dm:67]: doesn't look feasible

**Comment by itseasytosee** [comment]
> I have no real objections against this PR so I might as well merge it. I took the liberty to remove commas from the UI.

Thanks :)

**Comment by datorangebottle** [comment]
> Medical could use a form to fill out as a record of physical exams.

Or a form that confirms the patient has elected for an optional/experimental surgery.

**Comment by DrDiasyl** [comment]
POG POG POG POG

**Comment by Ironnhawk** [comment]
@twilightwanderer I cannot love you more for this project

**Comment by twilightwanderer** [comment]
> @twilightwanderer I cannot love you more for this project. Do you plan on adding any more of the old formatting options or Praying :) The better Field method

Do you mean a bureaucracy system with tags like [field], [logo], [date]?

**Comment by Ironnhawk** [comment]
> > @twilightwanderer I cannot love you more for this project. Do you plan on adding any more of the old formatting options or Praying :) The better Field method
> 
> Do you mean a bureaucracy system with tags like [field], [logo], [date]?

Yea!!!!!!! <3

**Comment by Ironnhawk** [comment]
> 

Oh ok! I still love the change! <3

**Pull request merged**
